### PR TITLE
[Qwen3-Next] Add cutedsl decode/mtp kernel with transposed ssm_state and prefill gluon kernel for blackwell.

### PR DESCRIPTION
**Pull request by Jon-WZQ**
## Motivation

For Qwen3-Next model, we observed current preill/decode kernels cannot fully utilize the hardware efficiency of Blackwell. Thus, we implement gluon based prefill kernels and cutedsl based decode kernels for better performance. Here is a blog about this PR. https://zhuanlan.zhihu.com/p/2003887397411258684

## Modifications

1. For prefill, we rewrite recompute_w_u_fwd_kernel/chunk_fwd_kernel_o kernels in gluon with triton>=3.6.0.  https://github.com/sgl-project/sglang/pull/17983
2. For decode, we use cutedsl to write the fuse_recurrent kernel for both decode/mtp. Besides, we transpose the shape of mamba_ssm_state from [B,H,K,V] to [B,H,V,K], which makes the memory contiguous in K dim. We achieve this only in ssm_state initialization phase and prefill kernel. It will introduce almost no additional overhead.

## Accuracy Tests

 We use `python python/sglang/jit_kernel/tests/test_cutedsl_gdn.py` for decode kernel accuracy and performance test.

**Decode Precision Test Results**

| B(BS) | kernel               | dtype | max_diff | mean_diff | fail_rate | check |
|------:|----------------------|:-----:|---------:|----------:|----------:|:-----:|
| 8     | cutedsl_gdn          | f32   | 0.00e+00 | 1.00e-09  | 0.00%     | ok    |
| 8     | cutedsl_gdn_transpose| bf16  | 6.10e-05 | 1.87e-09  | 0.00%     | ok    |
| 8     | cutedsl_gdn_transpose| f32   | 4.88e-04 | 1.49e-08  | 0.00%     | ok    |
| 16    | cutedsl_gdn          | f32   | 4.88e-04 | 1.26e-08  | 0.00%     | ok    |
| 16    | cutedsl_gdn_transpose| bf16  | 1.22e-04 | 1.93e-09  | 0.00%     | ok    |
| 16    | cutedsl_gdn_transpose| f32   | 2.44e-04 | 5.90e-09  | 0.00%     | ok    |
| 32    | cutedsl_gdn          | f32   | 9.77e-04 | 9.43e-09  | 0.00%     | ok    |
| 32    | cutedsl_gdn_transpose| bf16  | 6.10e-05 | 1.68e-09  | 0.00%     | ok    |
| 32    | cutedsl_gdn_transpose| f32   | 9.77e-04 | 1.03e-08  | 0.00%     | ok    |
| 64    | cutedsl_gdn          | f32   | 9.77e-04 | 9.51e-09  | 0.00%     | ok    |
| 64    | cutedsl_gdn_transpose| bf16  | 2.44e-04 | 3.82e-09  | 0.00%     | ok    |
| 64    | cutedsl_gdn_transpose| f32   | 4.88e-04 | 2.86e-09  | 0.00%     | ok    |
| 128   | cutedsl_gdn          | f32   | 4.88e-04 | 5.82e-09  | 0.00%     | ok    |
| 128   | cutedsl_gdn_transpose| bf16  | 4.88e-04 | 4.48e-09  | 0.00%     | ok    |
| 128   | cutedsl_gdn_transpose| f32   | 2.44e-04 | 2.98e-09  | 0.00%     | ok    |
| 256   | cutedsl_gdn          | f32   | 9.77e-04 | 4.75e-09  | 0.00%     | ok    |
| 256   | cutedsl_gdn_transpose| bf16  | 4.88e-04 | 5.01e-09  | 0.00%     | ok    |
| 256   | cutedsl_gdn_transpose| f32   | 9.77e-04 | 4.01e-09  | 0.00%     | ok    |

**MTP Decode Precision Test Results**

| T(BS) | max_diff | mean_diff | fail_rate | check |
|-----:|---------:|----------:|----------:|:-----:|
| 4    | 3.81e-06 | 1.56e-10  | 0.00%     | ok    |
| 4    | 1.53e-05 | 6.14e-10  | 0.00%     | ok    |
| 8    | 4.88e-04 | 9.79e-09  | 0.00%     | ok    |
| 8    | 9.77e-04 | 1.84e-08  | 0.00%     | ok    |
| 16   | 4.88e-04 | 5.67e-09  | 0.00%     | ok    |
| 16   | 4.88e-04 | 7.04e-09  | 0.00%     | ok    |
| 32   | 1.95e-03 | 1.61e-08  | 0.00%     | ok    |
| 32   | 1.25e-01 | 3.24e-07  | 0.00%     | ok    |
| 48   | 9.77e-04 | 1.17e-08  | 0.00%     | ok    |
| 48   | 3.91e-03 | 1.34e-08  | 0.00%     | ok    |

## Benchmarking and Profiling

**Performance Test Results (decode mode)**

head_q=head_k=16, head_v=32, dim=128

#### BF16 Results
| B(BS) | Triton (us) | CuTeDSL (us) | speedup | check |
|-----:|------------:|-------------:|--------:|:------|
| 8 | 9.95± 0.03 | 0.00± 0.00 | 0.00x | ok [cutedsl_gdn, bf16] |
| 8 | 9.95± 0.03 | 6.14± 0.02 | 1.62x | ok [cutedsl_gdn_transpose, bf16] |
| 16 | 16.92± 0.02 | 0.00± 0.00 | 0.00x | ok [cutedsl_gdn, bf16] |
| 16 | 16.92± 0.02 | 10.23± 0.01 | 1.65x | ok [cutedsl_gdn_transpose, bf16] |
| 32 | 31.53± 0.02 | 0.00± 0.00 | 0.00x | ok [cutedsl_gdn, bf16] |
| 32 | 31.53± 0.02 | 19.07± 0.01 | 1.65x | ok [cutedsl_gdn_transpose, bf16] |
| 64 | 60.11± 0.01 | 0.00± 0.00 | 0.00x | ok [cutedsl_gdn, bf16] |
| 64 | 60.11± 0.01 | 35.63± 0.02 | 1.69x | ok [cutedsl_gdn_transpose, bf16] |
| 128 | 119.16± 0.02 | 0.00± 0.00 | 0.00x | ok [cutedsl_gdn, bf16] |
| 128 | 119.16± 0.02 | 72.71± 0.02 | 1.64x | ok [cutedsl_gdn_transpose, bf16] |
| 256 | 234.96± 0.03 | 0.00± 0.00 | 0.00x | ok [cutedsl_gdn, bf16] |
| 256 | 234.96± 0.03 | 143.86± 0.04 | 1.63x | ok [cutedsl_gdn_transpose, bf16] |

#### FP32 Results
| B(BS) | Triton (us) | CuTeDSL (us) | speedup | check |
|-----:|:-------------:|:------------:|:-------:|:-------|
| 8 | 9.63± 0.01 | 9.47± 0.01 | 1.02x | ok [cutedsl_gdn, f32] |
| 8 | 9.63± 0.01 | 6.54± 0.01 | 1.47x | ok [cutedsl_gdn_transpose, f32] |
| 16 | 16.95± 0.03 | 16.64± 0.02 | 1.02x | ok [cutedsl_gdn, f32] |
| 16 | 16.95± 0.03 | 11.03± 0.02 | 1.54x | ok [cutedsl_gdn_transpose, f32] |
| 32 | 30.97± 0.03 | 26.27± 0.02 | 1.18x | ok [cutedsl_gdn, f32] |
| 32 | 30.97± 0.03 | 19.95± 0.01 | 1.55x | ok [cutedsl_gdn_transpose, f32] |
| 64 | 65.06± 0.05 | 54.31± 0.04 | 1.20x | ok [cutedsl_gdn, f32] |
| 64 | 65.06± 0.05 | 47.35± 0.03 | 1.37x | ok [cutedsl_gdn_transpose, f32] |
| 128 | 125.14± 0.44 | 102.60± 0.44 | 1.22x | ok [cutedsl_gdn, f32] |
| 128 | 125.14± 0.44 | 90.43± 0.13 | 1.38x | ok [cutedsl_gdn_transpose, f32] |
| 256 | 248.06± 2.39 | 200.63± 2.21 | 1.24x | ok [cutedsl_gdn, f32] |
| 256 | 248.06± 2.39 | 178.92± 0.80 | 1.39x | ok [cutedsl_gdn_transpose, f32] |

**MTP Performance Test Results (Draft Tokens 3)**

#### BF16

| T(BS) | Triton (us) | CuTeDSL (us) | Speedup |
|------:|:-----------:|:------------:|:-------:|
| 4 | 12.38± 0.07 | 9.59± 0.09 | 1.29x |
| 8 | 21.82± 0.10 | 15.79± 0.12 | 1.38x |
| 16 | 37.06± 0.10 | 24.69± 0.08 | 1.50x |
| 32 | 69.28± 0.10 | 44.49± 0.09 | 1.56x |
| 48 | 101.32± 0.10 | 64.72± 0.11 | 1.57x |

#### F32

| T(BS) | Triton (us) | CuTeDSL (us) | Speedup |
|------:|:-----------:|:------------:|:-------:|
| 4 | 12.92± 0.09 | 11.48± 0.10 | 1.13x |
| 8 | 22.85± 0.13 | 19.57± 0.17 | 1.17x |
| 16 | 41.20± 0.14 | 33.09± 0.13 | 1.24x |
| 32 | 73.65± 0.17 | 61.37± 0.11 | 1.20x |
| 48 | 108.95± 0.16 | 95.19± 0.12 | 1.14x |

**Qwen3-Next-80B-A3B-Instruct-FP8 E2E test**
`SGLANG_USE_CUTEDSL_GDN_DECODE_TRANSPOSE=1 python -m sglang.launch_server --model-path Qwen3-Next-80B-A3B-Instruct-FP8  --tp-size 2 --speculative-num-steps=2  --speculative-eagle-topk=1 --speculative-num-draft-tokens=3 --speculative-draft-model-path Qwen3-Next-80B-A3B-Instruct-FP8 --speculative-algorithm NEXTN`

***accuracy check***
python benchmark/gsm8k/bench_sglang.py --num-questions  128
Accuracy: 0.961
Invalid: 0.000
Latency: 10.909 s
Output throughput: 1834.008 token/s

***baseline vs cutedsl_gdn_transpose, In:Out=1:1500***
<table>
  <tr>
    <td>image</td>
    <td>image</td>
  </tr>
</table>

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
5. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 1]
```diff
@@ -820,12 +824,17 @@ def __init__(self, model_runner: ModelRunner):
...
 
         use_cutedsl = Envs.SGLANG_USE_CUTEDSL_GDN_DECODE.get()
-        rank0_log(f"CuTe DSL GDN decode enabled: {use_cutedsl}")
-        self._kernel_func = (
-            cutedsl_fused_sigmoid_gating_delta_rule_update
-            if use_cutedsl
-            else fused_sigmoid_gating_delta_rule_update
-        )
+        self.use_cutedsl_transpose = Envs.SGLANG_USE_CUTEDSL_GDN_DECODE_TRANSPOSE.get()
+        rank0_log(f"CuTe DSL GDN decode enabled: use_cutedsl: {use_cutedsl}, use_cutedsl_transpose: {self.use_cutedsl_transpose}")
```
Comment (BBuf) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: Should these two environment variables be checked for mutual exclusivity?

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py [hunk 2]
```diff
@@ -820,12 +824,17 @@ def __init__(self, model_runner: ModelRunner):
             ), f"{self.conv_states_shape[-1]=} should be less than {FLA_CHUNK_SIZE}"
 
         use_cutedsl = Envs.SGLANG_USE_CUTEDSL_GDN_DECODE.get()
-        rank0_log(f"CuTe DSL GDN decode enabled: {use_cutedsl}")
-        self._kernel_func = (
-            cutedsl_fused_sigmoid_gating_delta_rule_update
-            if use_cutedsl
-            else fused_sigmoid_gating_delta_rule_update
-        )
+        self.use_cutedsl_transpose = Envs.SGLANG_USE_CUTEDSL_GDN_DECODE_TRANSPOSE.get()
```
Comment (BBuf) [inline python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]:
```suggestion
        use_cutedsl_transpose = Envs.SGLANG_USE_CUTEDSL_GDN_DECODE_TRANSPOSE.get()
```

File: python/sglang/jit_kernel/cutedsl_gdn_transpose.py [hunk 3]
```diff
@@ -0,0 +1,1029 @@
...
+
+    if b is None:
+        b = torch.ones_like(q[..., 0])
+
+    if b.dim() == 2:
+        b = b.unsqueeze(0)
+    if a.dim() == 2:
+        a = a.unsqueeze(0)
+
+    q_ = from_dlpack(q.detach(), assumed_align=16)
```
Comment (BBuf) [inline python/sglang/jit_kernel/cutedsl_gdn_transpose.py:474]: Should we add an address align check?

**Review by BBuf — APPROVED** [review]

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 1 above)
Comment (Torches97) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: Yes, we have verified end-to-end precision across three configurations: the default kernel, `SGLANG_USE_CUTEDSL_GDN_DECODE`=1, and `SGLANG_USE_CUTEDSL_GDN_DECODE_TRANSPOSE`=1.

File: python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py (hunk 2 above)
Comment (Torches97) [reply python/sglang/srt/layers/attention/hybrid_linear_attn_backend.py]: We will use the env `SGLANG_USE_CUTEDSL_GDN_DECODE_TRANSPOSE` in forward_extend and forward_decode.

File: python/sglang/jit_kernel/cutedsl_gdn_transpose.py [hunk 4]
```diff
@@ -0,0 +1,1031 @@
...
+            mInter, (1, 1, 1, tiler_mn_h[0], tiler_mn_h[1])
+        )
+
+    B = mQ.shape[0] if cu_seqlens is None else cu_seqlens.shape[0] - 1
+    T = mK.shape[1]
+    HK = mK.shape[2]
+    HV = mV.shape[2]
+    blocks_per_head = mK.shape[-1] // BV
+
+    assert T // CACHE_STEPS == B, "batch * CACHE_STEPS must be equal to T"
```
Comment (yuan-luo) [inline python/sglang/jit_kernel/cutedsl_gdn_transpose.py:861]: If cache_steps not input, the cache_steps = T, this asserts fallback to T//T==B, namely B must be 1. That means by default bs > 1 will assert, since this kernel is for decode only, it might need to be addressed in function name, or add some comment.

File: python/sglang/jit_kernel/cutedsl_gdn_transpose.py [hunk 5]
```diff
@@ -0,0 +1,1031 @@
...
+    gK = cute.zipped_divide(mK, (1, 1, k_tiler_mn[0], k_tiler_mn[1]))
+    gV = cute.zipped_divide(mV, (1, 1, v_tiler_mn[0], v_tiler_mn[1]))
+    gO = cute.zipped_divide(mO, (1, 1, v_tiler_mn[0], v_tiler_mn[1]))
+    gH = cute.zipped_divide(mH, (1, 1, tiler_mn_h[0], tiler_mn_h[1]))
+    gB = cute.zipped_divide(mB, (1, 1, 1))
+    gA = mA
+    ga = cute.zipped_divide(ma, (1, 1, 1))
+    gdt_bias = mdt_bias
+
+    B = mQ.shape[0] if cu_seqlens is None else cu_seqlens.shape[0] - 1
```
Comment (yuan-luo) [inline python/sglang/jit_kernel/cutedsl_gdn_transpose.py:390]: [nit] In some places it uses `if cu_seqlens is None`, while in the same file L947 it uses `if cu_seqlens != None:`. Keeping the style unified is decent.

**Comment by Jon-WZQ** [comment]
In the latest commit, we tune y_threads to 4. Here is the latest results.

**MTP Performance Test Results (BF16, Draft Tokens 3)**

script: test_cutedsl_gdn.py
| T(BS) | Triton (us) | CuTeDSL (us) | Speedup |
|------:|:-----------:|:------------:|:-------:|
| 4 | 12.27± 0.09 | 9.52± 0.07 | 1.29x |
| 8 | 21.63± 0.10 | 15.58± 0.12 | 1.39x |
| 16 | 36.87± 0.12 | 22.24± 0.08 | 1.66x |
| 32 | 69.00± 0.10 | 38.74± 0.10 | 1.78x |
| 48 | 101.42± 0.12 | 56.28± 0.11 | 1.80x |

Besides, we also compare with the https://github.com/sgl-project/sglang/pull/19150 (Impressed Work!)

script: benchmark_gdn_transpose_vs_flashinfer.py
| BS | CuteDSL (us) | FI (us) https://github.com/sgl-project/sglang/pull/19150  | Speedup |
|------:|:-----------:|:------------:|:-------:|
| 1 | 6.12 | 7.96 | 1.3x |
| 32 | 98.99 | 114.33 | 1.16x |
| 64 | 189.08 | 209.70 | 1.10x |
| 128 | 375.89 | 408.63 | 1.08x |

cc @hlu1 @yuan-luo

File: python/sglang/jit_kernel/cutedsl_gdn_transpose.py (hunk 4 above)
Comment (Torches97) [reply python/sglang/jit_kernel/cutedsl_gdn_transpose.py:861]: OK, we will add some comments.
